instant-xml/src/parse.rs

```diff
@@ -1,48 +1,154 @@
+                        // Other attributes
+                        attributes.push((local.to_string(), value.to_string()));
+                    } else {
+                        // TODO: Can the attributes have the prefix?
+                        todo!();
+                    }
+                }
+                Ok(Token::Text { text }) => {
+                    return Ok(Some(XmlRecord::Element(text.to_string())));
+                }
+                Ok(_) => return Err(Error::UnexpectedToken),
+                Err(e) => return Err(Error::Parse(e)),
             }
-            Some(Ok(_)) => Err(Error::UnexpectedValue),
-            Some(Err(err)) => Err(err.into()),
-            None => Err(Error::UnexpectedEndOfStream),
         }
     }
 
-    fn element_end(self, _: Option<&str>, _: &str) -> Result<(), Error> {
-        match self {
-            Some(Ok(Token::ElementEnd { end, .. })) => match end {
-                ElementEnd::Open => todo!(),
-                ElementEnd::Close(_, _) => todo!(),
-                ElementEnd::Empty => Ok(()),
-            },
-            Some(Ok(_)) => Err(Error::UnexpectedValue),
-            Some(Err(err)) => Err(err.into()),
-            None => Err(Error::UnexpectedEndOfStream),
+    pub fn peek_next_tag(&mut self) -> Result<Option<XmlRecord>> {
+        let item = match self.internal_iter.peek() {
+            Some(v) => v,
+            None => return Ok(None),
+        };
+
+        println!("peek: {:?}", &item);
+        match item {
+            Ok(Token::ElementStart { prefix, local, .. }) => {
+                let prefix_ret = if prefix.is_empty() {
+                    None
+                } else {
+                    Some(prefix.to_string())
+                };
+
+                Ok(Some(XmlRecord::Open(TagData {
+                    key: local.to_string(),
+                    attributes: Vec::new(),
+                    default_namespace: None,
+                    namespaces: None,
+                    prefix: prefix_ret,
+                })))
+            }
+            Ok(Token::ElementEnd { end, .. }) => {
+                if let ElementEnd::Close(..) = end {
+                    if self.stack.is_empty() {
+                        return Err(Error::UnexpectedEndOfStream);
+                    }
+
+                    return Ok(Some(XmlRecord::Close(
+                        self.stack.last().unwrap().to_string(),
+                    )));
+                }
+                Err(Error::UnexpectedToken)
+            }
+            Ok(_) => Err(Error::UnexpectedToken),
+            Err(e) => Err(Error::Parse(*e)),
         }
     }
 }
 
-pub trait Parse {
-    fn element_start(self, ns: Option<&str>, tag: &str) -> Result<(), Error>;
-    fn element_end(self, ns: Option<&str>, tag: &str) -> Result<(), Error>;
+impl<'a> Iterator for XmlParser<'a> {
+    type Item = Result<XmlRecord>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.parse_next() {
```

Comment:
Any reason not to just inline the implementation of `parse_next()` here?

Reply:
Inlined